### PR TITLE
cni/linux: ensure only the latest sandbox has external-ids:iface-id set

### PR DESCRIPTION
**Pull request by dcbw**
Kubelet may keep multiple sandboxes running for a given pod if some are
waiting for garbage collection. We need to make sure that only the
latest sandbox has iface-id set to the container's namespace/name or
ovn-controller may get confused about which OVS port to associate with
the Pod's logical switch port.

@squeed @girishmg @danwinship

### CODE REVIEW

**Comment by danwinship** [comment]
Shouldn't the old interface have been deleted by this point? Is this a sign that some previous cleanup failed?

Do you have logs showing this happening?

**Comment by squeed** [comment]
In the case where the sandbox dies, there's no guarantee that the previous one will have a CNI DEL issued before the replacement has a CNI ADD. I've personally seen this, and it's pretty easy to reproduce: just kill the pause container.

**Comment by squeed** [comment]
In the case where the sandbox dies, there's no guarantee that the previous one will have a CNI DEL issued before the replacement has a CNI ADD. I've personally seen this, and it's pretty easy to reproduce: just kill the pause container.

**Comment by dcbw** [comment]
> Shouldn't the old interface have been deleted by this point? Is this a sign that some previous cleanup failed?
> 
> Do you have logs showing this happening?

@danwinship to echo what Casey said, Kubelet may keep a couple sandboxes for a given pod around up until the GC interval. Then it will delete them as they expire. Only one of the sandboxes will be "live" (eg SANDBOX_READY) but for OVN only that sandbox should have the iface-id set.

**Comment by squeed** [comment]
Without knowing the OVN data model very well, is it possible to use the containerid, which *will* be unique, as the externalid?

**Comment by girishmg** [comment]
@squeed 

The `iface-id` is already unique across multiple instances of these sandboxes for a a given pod. I think the issue is we cannot have multiple physical OVS internal ports that are all tagged with the same 'iface-id'. The `ovn-controller` will be confused as to which of the physical port it needs to bind the logical flows to.

File: go-controller/pkg/cni/helper_linux.go [hunk 1]
```diff
@@ -266,6 +266,17 @@ func (pr *PodRequest) ConfigureInterface(namespace string, podName string, ifInf
 
 	ifaceID := fmt.Sprintf("%s_%s", namespace, podName)
 
+	// Find and remove any existing OVS port with this iface-id. Pods can
+	// have multiple sandboxes if some are waiting for garbage collection,
+	// but only the latest one should have the iface-id set.
+	uuids, _ := ovsFind("Interface", "_uuid", "external-ids:iface-id="+ifaceID)
+	for _, uuid := range uuids {
+		if out, err := ovsExec("remove", "Interface", uuid, "external-ids", "iface-id"); err != nil {
```
Comment (girishmg) [inline go-controller/pkg/cni/helper_linux.go]: @dcbw if you want to remote the interface itself, then you will need "destroy".

```
# ovs-vsctl --help |grep 'destroy TBL'
  destroy TBL REC             delete RECord from TBL
```

Comment (dcbw) [reply go-controller/pkg/cni/helper_linux.go]: @girishmg I wasn't actually intending to destroy the Interface object, just to clear its iface-id from its Interface record. The actual Interface would be destroyed when kubelet/CRI finally called CNI DEL for the sandbox.

**Comment by dcbw** [comment]
> Without knowing the OVN data model very well, is it possible to use the containerid, which _will_ be unique, as the externalid?

@squeed the master doesn't have any access to sandbox-level stuff since that's all done on the node and not recorded in the Kube API AFAIK...

> The `iface-id` is already unique across multiple instances of these sandboxes for a a given pod. I think the issue is we cannot have multiple physical OVS internal ports that are all tagged with the same 'iface-id'. The `ovn-controller` will be confused as to which of the physical port it needs to bind the logical flows to.

@girishmg I think what Casey is saying is that we can avoid this problem entirely if we name the OVN port after the CNI sandbox ID (which will be unique) rather than the pod namespace+name (which is shared between sandboxes). However, we can't do that because the master has no idea what the sandbox is.

**Pull request merged**